### ISSUE TITLE
Adding metrics

### ISSUE THREAD

**Issue by JoyHighmore** [ashawkey/torch-ngp#25]
Hi!

I'm trying to add some metrics into the code. Could you please give a clue about how to realize that?

For me, it always has some bug saying there is no attribute "clear" etc in my function object.

Sincerely

**Comment by ashawkey** [comment]
Hi,
If I understand correctly, you are trying to use the `metrics` for `Trainer`? It is an old piece of code I borrowed from my other projects, so it may be confusing to use. 

An example of a metric class can be:
```python
class ROCMeter:
    def __init__(self):
        self.truths = []
        self.preds = []

    def clear(self):
        self.truths = []
        self.preds = []

    def prepare_inputs(self, *inputs):
        outputs = []
        for i, inp in enumerate(inputs):
            if torch.is_tensor(inp):
                inp = inp.detach().cpu().numpy()
            outputs.append(inp)

        return outputs

    def update(self, preds, truths):
        # [B, ], [B, ]
        preds, truths = self.prepare_inputs(preds, truths)
        self.truths.append(truths)
        self.preds.append(preds)


    def measure(self):
        return roc_auc_score(np.concatenate(self.truths, 0), np.concatenate(self.preds, 0))

    def write(self, writer, global_step, prefix=""):
        writer.add_scalar(os.path.join(prefix, "roc"), self.measure(), global_step)

    def report(self):
        return f'ROC = {self.measure():.6f}'
```

and used like:

```python
trainer = Trainer(..., metrics=[ROCMeter(),], ...)
```

I will add a PSNR metric recently.

**Comment by JoyHighmore** [comment]
Thank you sooo much! I'll try to implement that. Moreover, can you explain in SDF why are you using the negative sign here?         sdfs[self.num_samples // 2:] = -self.sdf_fn(points[self.num_samples // 2:])[:,None].astype(np.float32)

**Comment by ashawkey** [comment]
@JoyHighmore I have added the PSNR metric, you can try now!
As for the sdf value, it is just a convention (inner is negative, outer is positive), without the negation it should also work since we extract surface at `sdf == 0`.

(Besides, I find that the PSNR for lego is still worse compared to the original implementation, and I'll try to figure it out. If you find any problem, welcome to report!)

**Comment by JoyHighmore** [comment]
Sure, there is one more thing that I'd like to discuss. I've noticed that there are some code trying to calculate the sdf by igl.signed_distance. I've also implement both igl and pysdf. It is found that they had quite different results, which not only with the different values but also different indications for inside and outside categories. Do you have any thoughts? What's your consideration for the pysdf? Also, are you trying to have balanced positive and negative sdf since you are using self.num_samples // 2 in this equation?

Best,
Joy

**Comment by ashawkey** [comment]
@JoyHighmore That's strange, the SDF value should be unique as long as the mesh is fixed. Inside/Outside doesn't really matter as the only difference is a negation. 
In fact the pos/neg points are not balanced, I followed the original implementation to  sample 1/8 random points, 3/8 surface points, and 4/8 perturbed surface points.

**Comment by ashawkey** [comment]
closed for now.
